dvc/remote/base.py

```diff
@@ -142,23 +149,32 @@ def cache(self):
     def get_file_checksum(self, path_info):
         raise NotImplementedError
 
-    def _collect_dir(self, path_info):
-        dir_info = {}
+    def _get_file_checksum_future(self, file_info, executor):
+        checksum = self.state.get(file_info)
+        if checksum:
+            checksum = as_future(checksum)
+        else:
+            checksum = executor.submit(self.get_file_checksum, file_info)
```

Comment:
Why not just submit a `_get_file_checksum` to the executor that would decide by itself if it wants to use state or actually compute checksum?

Reply:
Also, it is not relevant right now since we don't implement batching yet, but in the future I guess we would've tried to retrieve checksums for all of these files from state _before_ submitting anything to the executor, right? That might be useful to keep in mind and maybe that will make it worth it to do the same here, even though without batching yet.

Reply:
First  comment will be hard to implement, as sqlite by default does not support operating in multiple threads. We can achieve that by setting `check_same_thread` in sqlite3.connect to False, but that will result in random tests fail (checked it), probably because of timestamp updates in `get` method.

I believe that second is premature optimization, though I agree that we will have to introduce that at some point, maybe it will be for the best to do it now.

Reply:
Great points! My thinking with the second one was that it would at least get rid of wrapping things into as_future and will provide a nice base for the future. But ok, lets go with the current approach for now.

Reply:
@efiop new change does not include as_future, and it retrieves checksums before scheduling md5 execution. I guess we can hit 2 birds with this stone (if it passes review, of course)